extensions/amp-viewer-assistance/0.1/amp-viewer-assistance.js

```diff
@@ -0,0 +1,190 @@
+/**
+ * Copyright 2019 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {ActionTrust} from '../../../src/action-constants';
+import {Services} from '../../../src/services';
+import {dev, user} from '../../../src/log';
+import {dict} from '../../../src/utils/object';
+import {isExperimentOn} from '../../../src/experiments';
+import {tryParseJson} from '../../../src/json';
+
+
+/** @const {string} */
+const TAG = 'amp-viewer-assistance';
+
+/** @const {string} */
+const GSI_TOKEN_PROVIDER = 'actions-on-google-gsi';
+
+export class AmpViewerAssistance {
+  /**
+   * @param {!../../../src/service/ampdoc-impl.AmpDoc} ampdoc
+   */
+  constructor(ampdoc) {
+    const assistanceElement = ampdoc.getElementById('amp-viewer-assistance');
+
+    /** @private {boolean} */
+    this.enabled_ = !!assistanceElement;
+    if (!this.enabled_) {
+      return;
+    }
+
+    /** @const @private */
+    this.ampdoc_ = ampdoc;
+
+    /** @const @private {!Element} */
+    this.assistanceElement_ = dev().assertElement(assistanceElement);
+
+    /** @const @private {JsonObject|null|undefined} */
+    this.configJson_ = tryParseJson(this.assistanceElement_.textContent, e => {
+      throw user().createError(
+          'Failed to parse "amp-viewer-assistance" JSON: ' + e);
+    });
+
+    /** @private @const {!../../../src/service/viewer-impl.Viewer} */
+    this.viewer_ = Services.viewerForDoc(ampdoc);
+
+    /** @private @const {!../../../src/service/action-impl.ActionService} */
+    this.action_ = Services.actionServiceForDoc(this.assistanceElement_);
+
+    /** @private @const {!../../../src/service/vsync-impl.Vsync} */
+    this.vsync_ = Services.vsyncFor(ampdoc.win);
+  }
+
+  /**
+   * @param {!../../../src/service/action-impl.ActionInvocation} invocation
+   * @return {?Promise}
+   * @private
+   */
+  actionHandler_(invocation) {
```

Comment:
Sorry didn't catch this earlier. You should actually use `this.registerAction()` here instead. See `base-element.js#registerAction`.

Then you can also remove the whitelisting in `presubmit-checks.js`.

Reply:
As we discussed offline, we are following the pattern set by [amp-access](https://github.com/ampproject/amphtml/blob/b1993f31d96bd7c287f7a7e47b8c9e2534def2d8/extensions/amp-access/0.1/amp-access.js#L630) to justify using our own action handler since we are using a `<script>` tag instead of an `<amp-viewer-assistance>` element.